src/crypto.rs

```diff
@@ -0,0 +1,206 @@
+    }
+}
+
+/// Implementation of GCM/CCM set/get properties for encrypt/seal
+pub trait ContextWithXcmEncryptProperties: Context {
+    /// Sets the GCM AAD property.
+    fn set_property_gcm_aad(&self, gcm_aad: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmAad, gcm_aad)
+    }
+    /// Sets the GCM tag length property.
+    fn set_property_gcm_tag_len(&self, gcm_tag_len: usize) -> Result<()>
+    {
+        context_set_property_single(self, types::Property::GcmTagLen, gcm_tag_len)
+    }
+    /// Sets the CCM AAD property.
+    fn set_property_ccm_aad(&self, ccm_aad: &[u8], input_len: usize) -> Result<()>;
+    // {
+    //     OPERATION_set_input_length(self, input_len)?;
+    //     crypto::context_set_property_multiple(ctx, types::Property::CcmAad, ccm_aad)
+    // }
+    /// Sets the CCM tag length property.
+    fn set_property_ccm_tag_len(&self, ccm_tag_len: usize) -> Result<()>
+    {
+        context_set_property_single(self, types::Property::CcmTagLen, ccm_tag_len)
+    }
+    /// Returns the GCM tag property.
+    fn get_property_gcm_tag(&self) -> Result<Vec<u8>>
+    {
+        context_get_property_multiple(self, types::Property::GcmTag)
+    }
+    /// Returns the CCM tag property.
+    fn get_property_ccm_tag(&self) -> Result<Vec<u8>>
+    {
+        context_get_property_multiple(self, types::Property::CcmTag)
+    }
+}
+
+/// Implementation of all GCM/CCM properties for decrypt/open
+pub trait ContextWithXcmDecryptProperties: Context {
+    /// Sets the GCM AAD property.
+    fn set_property_gcm_aad(&self, gcm_aad: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmAad, gcm_aad)
+    }
+    /// Sets the GCM tag property.
+    fn set_property_gcm_tag(&self, gcm_tag: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmTag, gcm_tag)
+    }
+    /// Sets the CCM AAD property.
+    fn set_property_ccm_aad(&self, ccm_aad: &[u8], input_len: usize) -> Result<()>;
+    // {
+    //     OPERATION_set_input_length(self, input_len)?;
+    //     crypto::context_set_property_multiple(ctx, types::Property::CcmAad, ccm_aad)
+    // }
+    /// Sets the CCM tag property.
+    fn set_property_ccm_tag(&self, ccm_tag: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::CcmTag, ccm_tag)
+    }
+}
+
+
+/// Initializes the library. Must be called before any other crypto
+/// function. Should be called once in each thread that uses yaca.
+pub fn initialize() -> Result<()>
+{
+    let r = unsafe {
+        lib::yaca_initialize()
+    };
+    conv::res_c_to_rs(r)
+}
+
+/// Cleans up the library.
+/// Must be called before exiting the thread that called yaca_initialize().
+pub fn cleanup()
+{
+    unsafe {
+        lib::yaca_cleanup()
+    }
+}
+
+/// Safely compares first length bytes of two buffers.
+pub fn memcmp(first: &[u8], second: &[u8], length: usize) -> Result<bool>
```

Comment:
This is the question to those crypto-literate. As far as I remember there was something more safe about this one than any other memcmp. To the point, that OpenSSL recommended using it instead of system ones. We also refer to this function in sign (I've added docs in the next patchset, you can see that).

I'd leave it. 

About length, it allows to have length < min. Is there a reason to do so? I don't know. It doesn't hurt I guess. I changed panic to cutting the value though. So you can pass std::usize::MAX safely.

Reply:
Apparently memcmp can "leak timing information" so a crappy one is provided as a replacement, on purpose: https://linux-crypto.vger.kernel.narkive.com/eEfuKzM5/patch-crypto-memcmp-add-constant-time-memcmp

Go figure...

Yeah, let's leave it... Love crypto, can't get enough of it....